lib/src/client_offline_mixin.dart

```diff
@@ -0,0 +1,662 @@
+    void handleConnectivityResult(ConnectivityResult result) {
+      isOnline.value = resultIsOnline(result);
+    }
+
+    final result = await Connectivity().checkConnectivity();
+    handleConnectivityResult(result);
+    if (isOnline.value) {
+      // wifi or mobile is connected, but double check internet connectivity
+      await checkOnlineStatus();
+    }
+    Connectivity().onConnectivityChanged.listen(handleConnectivityResult);
+  }
+
+  Future<Response> handleOfflineRequest({
+    required Uri uri,
+    required HttpMethod method,
+    required Future<Response<dynamic>> Function(HttpMethod,
+            {String path,
+            Map<String, String> headers,
+            Map<String, dynamic> params,
+            ResponseType? responseType,
+            String cacheModel,
+            String cacheKey,
+            String cacheResponseIdKey,
+            String cacheResponseContainerKey})
+        call,
+    String path = '',
+    Map<String, String> headers = const {},
+    Map<String, dynamic> params = const {},
+    ResponseType? responseType,
+    String cacheModel = '',
+    String cacheKey = '',
+    String cacheResponseIdKey = '',
+    String cacheResponseContainerKey = '',
+    Map<String, Object?>? previous,
+  }) async {
+    if (!headers.containsKey('X-Appwrite-Timestamp')) {
+      headers['X-Appwrite-Timestamp'] =
+          DateTime.now().toUtc().toIso8601String();
+    }
+    final pathSegments = uri.pathSegments;
+    String queuedWriteKey = '';
+
+    final store = getModelStore(cacheModel);
+    switch (method) {
+      case HttpMethod.get:
+        if (cacheKey.isNotEmpty) {
+          final recordRef = store.record(cacheKey);
+          final record = await recordRef.get(db);
+          if (record != null) {
+            updateAccessedAt(db, store.name, cacheKey);
+            return Response(data: record);
+          }
+        } else {
+          final finder = Finder(limit: 25);
+          // TODO: await both at same time
+          final records = await store.find(db, finder: finder);
+          db.transaction((txn) async {
+            for (final record in records) {
+              await updateAccessedAt(txn, store.name, record.key);
+            }
+          });
+          final count = await store.count(db);
+          return Response(data: {
+            'total': count,
+            cacheResponseContainerKey: records.map((record) {
+              final map = Map<String, dynamic>();
+              record.value.entries.forEach((entry) {
+                map[entry.key] = entry.value;
+              });
+              return map;
+            }).toList(),
+          });
+        }
+        throw AppwriteException(
+          "Client is offline and data is not cached",
+          0,
+          "general_offline",
+        );
+      case HttpMethod.post:
+      case HttpMethod.patch:
+      case HttpMethod.put:
+      case HttpMethod.delete:
+        switch (method) {
+          case HttpMethod.post:
```

Comment:
I think so, yes, because the code starting from `final completer = Completer<Response>();` is the same for all these HTTP methods.

Reply:
Could that be moved to it's own method and called from the end of each case?

Reply:
@abnegate, I'd have to create a function with a pretty nasty signature:

```dart
  Future<Response<dynamic>> completeQueuedWrite({
    required String queuedWriteKey,
    required Future<Response<dynamic>> Function(HttpMethod,
            {String path,
            Map<String, String> headers,
            Map<String, dynamic> params,
            ResponseType? responseType,
            String cacheModel,
            String cacheKey,
            String cacheResponseIdKey,
            String cacheResponseContainerKey})
        call,
    required HttpMethod method,
    String path = '',
    Map<String, String> headers = const {},
    Map<String, dynamic> params = const {},
    ResponseType? responseType,
    required StoreRef<String, Map<String, Object?>> store,
    required String cacheKey,
  }) {
    final completer = Completer<Response>();
    // Declare listener first so it can be referenced inside itself
    Function() listener = () {};
    listener = () async {
      while (true) {
        final queuedWrites = await listQueuedWrites(db);

        if (queuedWrites.isEmpty) {
          break;
        }

        if (queuedWrites.first.key != queuedWriteKey) {
          await Future.delayed(Duration.zero);
          continue;
        }

        try {
          final res = await call(
            method,
            headers: headers,
            params: params,
            path: path,
            responseType: responseType,
          );

          await db.transaction((txn) async {
            final futures = <Future>[];
            if (method == HttpMethod.post) {
              futures.add(upsertCache(txn, store, res.data, key: cacheKey));
            }

            futures.add(deleteQueuedWrite(txn, queuedWriteKey));

            await Future.wait(futures);
          });

          completer.complete(res);
        } on AppwriteException catch (e) {
          if (e.message == "Bad state: Can't finalize a finalized Request.") {
            continue;
          }
          if (!completer.isCompleted) {
            if (e.code == 404) {
              // delete from cache
              await db.transaction((txn) async {
                await deleteCache(txn, store, key: cacheKey);
                await deleteQueuedWrite(txn, queuedWriteKey);
              });
            } else if ((e.code ?? 0) >= 400) {
              // restore cache
              final previous =
                  queuedWrites.first.value['previous'] as Map<String, Object?>?;
              await db.transaction((txn) async {
                if (previous != null) {
                  await upsertCache(txn, store, previous, key: cacheKey);
                }
                await deleteQueuedWrite(txn, queuedWriteKey);
              });
            }
            completer.completeError(e);
          }
        } catch (e) {
          if (!completer.isCompleted) {
            // restore cache
            final previous =
                queuedWrites.first.value['previous'] as Map<String, Object?>?;
            if (previous != null) {
              await db.transaction((txn) async {
                await upsertCache(txn, store, previous, key: cacheKey);
                await deleteQueuedWrite(txn, queuedWriteKey);
              });
            }
            completer.completeError(e);
          }
        }
        isOnline.removeListener(listener);
        break;
      }
    };
    isOnline.addListener(listener);
    return completer.future;
  }
```

Is this okay?

Reply:
Hmm it's not ideal, what about first checking `if` the method is `get`, and since that either returns or throws, the switch can trail it instead of being nested?

Reply:
@abnegate, updated! How is everything now?